### PR TITLE
Add maxWorkgroupParallelDims option to TileAndDistributeToWorkgroupsPass

### PR DESCRIPTION
**Pull request by KoolJBlack**
`maxWorkgroupParallelDims` option default to 3 which will distribute workgroups along `X`, `Y`, and `Z` dimensions (current behavior we have now). 

Setting the `maxWorkgroupParallelDims` option to 1 will dispatch all workgroups along the `X` dimension only. 

Addresses #12642. 

Note: option has not been enabled for the CUDA path yet.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp [hunk 1]
```diff
@@ -237,6 +239,8 @@ struct LowerDispatchWorkgroupCountForDagRootOp
 
   /// Loops that are partitioned.
   SmallVector<unsigned> partitionedLoops;
+
+  int32_t maxWorkgroupParallelDims;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:250]: Please add comments.

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp [hunk 2]
```diff
@@ -302,24 +306,42 @@ struct LowerDispatchWorkgroupCountFromSetEncodingOp
...
+  TileAndDistributeToWorkgroupsPass(int32_t maxWorkgroupParallelDims)
+      : maxWorkgroupParallelDims(maxWorkgroupParallelDims) {}
   void getDependentDialects(DialectRegistry &registry) const override {
     registry
         .insert<AffineDialect, IREE::Flow::FlowDialect, IREE::HAL::HALDialect,
                 linalg::LinalgDialect, IREE::LinalgExt::IREELinalgExtDialect,
                 scf::SCFDialect, tensor::TensorDialect>();
   }
 
+  void initOptions() {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: This is non-standard way of doing this, either do what is done here

https://github.com/openxla/iree/blob/663f92cd71ed3a85ca677e847d2446ff48423080/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp#L782

(there are other ways too, but this is what I can point to immediately.

File: compiler/src/iree/compiler/Codegen/Passes.h [hunk 3]
```diff
@@ -150,7 +150,7 @@ createTestPartitionableLoopsInterfacePass();
 
 /// Pass to tile and distribute to workgroups.
 std::unique_ptr<OperationPass<IREE::HAL::ExecutableVariantOp>>
-createTileAndDistributeToWorkgroupsPass();
+createTileAndDistributeToWorkgroupsPass(int32_t maxWorkgroupParallelDims = 3);
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.h]: Please dont hard code to `3`... use `kNumMaxParallelDims`.

File: compiler/src/iree/compiler/Codegen/Passes.td [hunk 4]
```diff
@@ -169,6 +169,11 @@ def TileAndDistributeToWorkgroups :
     Pass<"iree-codegen-tile-and-distribute-to-workgroups", "IREE::HAL::ExecutableVariantOp"> {
   let summary = "Tile and distribute operations to workgroups";
   let constructor = "mlir::iree_compiler::createTileAndDistributeToWorkgroupsPass()";
+  let options = [
+    Option<"maxWorkgroupParallelDims", "max-workgroup-parallel-dims", "int32_t",
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.td:173]: Please dont hard code to 3... use kNumMaxParallelDims.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Looks good! Just a few minor comments.

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: `kNumMaxParallelDims` is a constexpr constant. Can I reference that in TD files?

File: compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_max_workgroups.mlir [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// RUN: iree-opt --pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-codegen-tile-and-distribute-to-workgroups{max-workgroup-parallel-dims=1})), canonicalize, cse)' --split-input-file %s | FileCheck %s
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_max_workgroups.mlir]: nit: you could add this directly in `tile_and_distribute_to_workgroups.mlir` with a different prefix instead of having two files for this

**Review by ThomasRaoux — COMMENTED** [review]
Looks good, added a nit comment

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: If it is defined before the place the generated header file is included, then you can

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp (hunk 1 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:250]: I removed this variable.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:250]: The variable is still here.....

File: compiler/src/iree/compiler/Codegen/Passes.h (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Passes.h]: It is still hard-coded to 3....

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: I dont know what happens if the default isnt set...

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/Passes.h (hunk 3 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Passes.h]: fixed for real

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp (hunk 1 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:250]: eeh, yeah not sure what happened in my brain. pinky swear its commented this time.

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: I believe the Options are like `Optional` types, hence the `hasValue()` method. If I set this to empty, it uses the default value in the constructor `createTileAndDistributeToWorkgroupsPass()`. 

As stated earlier, I could not figure out how to get the `kNumMaxParallelDims` constant here. Its defined in Codgen/Utils.h, but the generated `compiler/src/iree/compiler/Codegen/Passes.h.inc` does not have any `#includes` in it and I cannot seem to add it. 

I couldn't find another example of using a constexpr in a TD file. Everything else uses literals. Any ideas?

File: compiler/src/iree/compiler/Codegen/SPIRV/Passes.cpp [hunk 6]
```diff
@@ -118,7 +118,8 @@ static void addBufferizePasses(OpPassManager &passManager,
 static void addTileAndDistributeToWorkgroupsPasses(
     OpPassManager &passManager, bool useFuseTensorPadWithConsumerPass = false,
     bool useWARForCooperativeMatrixCodegen = false) {
-  passManager.addPass(createTileAndDistributeToWorkgroupsPass());
+  passManager.addPass(
+      createTileAndDistributeToWorkgroupsPass(kNumMaxParallelDims));
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/SPIRV/Passes.cpp]: why is this needed? This is already the default value?

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: it should work but if you wanted to use `kNumMaxParallelDims` you just need to make sure it is part of a header included before `compiler/src/iree/compiler/Codegen/Passes.h.inc` (for instance you would need to make sure this gets included in `compiler/src/iree/compiler/Codegen/Passes.h`)

**Review by ThomasRaoux — APPROVED** [review]
Looks good to me once the last couple comments are addressed

File: compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir [hunk 7]
```diff
@@ -1,4 +1,5 @@
 // RUN: iree-opt --pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-codegen-tile-and-distribute-to-workgroups)), canonicalize, cse)' --split-input-file %s | FileCheck %s
+// RUN: iree-opt --pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-codegen-tile-and-distribute-to-workgroups{max-workgroup-parallel-dims=1})), canonicalize, cse)' --split-input-file %s | FileCheck %s -check-prefix=CHECKW
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir:2]: Just FYI, adding it this way means that all the tests in the file are run once more, but only one of these is checked. I think thats OK, it still shows that these tests run and dont crash with this option which is good.

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 4 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Passes.td:173]: Looked through how to not hard-coded... too complicated to manage the headers. This is just the default for command line option I guess. Please just hard-code it to 3 for now....

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir (hunk 7 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir:2]: IC makes sense. I originally had a separate file but Thomas suggested they be merged.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir:2]: what would be nice would be to re-use directly an existing test but that's a detail

File: compiler/src/iree/compiler/Codegen/SPIRV/Passes.cpp (hunk 6 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/SPIRV/Passes.cpp]: I removed them all

File: compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir (hunk 7 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir:2]: No need to change it. Just pointing out what is happening

Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Common/test/tile_and_distribute_to_workgroups.mlir:2]: sg i'll keep it as is.
